In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import os
import ast
import csv
import time
import matplotlib.ticker as mtick

# 원하시는 한글 폰트 이름을 지정하세요. (예: 'Malgun Gothic', 'AppleGothic')
plt.rc('font', family='Malgun Gothic') # Windows 환경
# 폰트 설정 후에는 음수 부호가 깨지는 것을 방지하기 위해 설정합니다.
plt.rcParams['axes.unicode_minus'] = False

# 유저 이벤트 데이터.csv 파일 불러오기

## hackle_properties

In [2]:
# csv 파일 불러오기
hackle_properties_path = r'C:\Users\USER\Documents\codeit_study\codeit_study\HYPE 프로젝트\hackle_properties.csv'
hackle_properties_column_names = ['id', 'session_id', 'user_id', 'language', 'osname', 'osversion', 'versionname', 'device_id']
hackle_properties_df = pd.read_csv(hackle_properties_path, header=None, names=hackle_properties_column_names)

In [3]:
hackle_properties_df.head()

id                            session_id  user_id language   osname  \
0   1          4OzYh3seq3VKytpSn5pvQkZNQii1      NaN    ko-KR      iOS   
1   2          8QXy31PQxbW9qLzq0Y1dhR8Ypm52  1046711    ko-KR      iOS   
2   3  6bcea65d-9f40-46fc-888c-700fe707483f  1545130       ko  Android   
3   4          XVYNT6zfhFWqIg9omwg2AHDjTLx2  1224793       ko  Android   
4   5          XFB2SPiGfjbVhvJ3Q3DBsaT3m2B3  1329450    ko-US      iOS   

  osversion versionname                             device_id  
0      16.0       2.0.0  590E7C79-CBA0-44D8-8BE3-3C9BFABBBC74  
1    16.5.1       2.0.3  D5417226-F71B-4A9E-A180-CD072F2AB279  
2        13       2.0.5  6bcea65d-9f40-46fc-888c-700fe707483f  
3        13       2.0.5  a05c1595-3e05-434b-8684-218b528bd725  
4    16.5.1       2.0.5  EAC6C0B3-7CE8-40EA-8A91-9977C0BA5EF3

In [4]:
hackle_properties_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525350 entries, 0 to 525349
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   id           525350 non-null  int64 
 1   session_id   525350 non-null  object
 2   user_id      443095 non-null  object
 3   language     525350 non-null  object
 4   osname       525350 non-null  object
 5   osversion    525350 non-null  object
 6   versionname  525350 non-null  object
 7   device_id    525350 non-null  object
dtypes: int64(1), object(7)
memory usage: 32.1+ MB


In [5]:
# 결측치 확인
hackle_properties_df.isnull().sum()

id                 0
session_id         0
user_id        82255
language           0
osname             0
osversion          0
versionname        0
device_id          0
dtype: int64

In [6]:
# user_id가 null인 데이터 -> 로그인 하지 않은 유저들에 대한 정보라 분석하기 어려움 -> 삭제
hackle_properties_cleaned = hackle_properties_df.copy()
hackle_properties_cleaned.dropna(subset=['user_id'], inplace=True)

In [7]:
hackle_properties_cleaned.isnull().sum()

id             0
session_id     0
user_id        0
language       0
osname         0
osversion      0
versionname    0
device_id      0
dtype: int64

In [8]:
# 데이터 중복 확인
hackle_properties_cleaned[hackle_properties_cleaned.duplicated()]

Empty DataFrame
Columns: [id, session_id, user_id, language, osname, osversion, versionname, device_id]
Index: []

## device_properties

In [9]:
# csv 파일 불러오기
device_properties_path = r'C:\Users\USER\Documents\codeit_study\codeit_study\HYPE 프로젝트\device_properties.csv'
device_properties_column_names = ['id', 'device_id', 'device_model', 'device_vendor']
device_properties_df = pd.read_csv(device_properties_path, header=None, names=device_properties_column_names)

In [10]:
device_properties_df.head()

id                             device_id device_model device_vendor
0   1  000007C9-E103-4EB5-9777-A9084D4952DF   iPhone14,7         Apple
1   2  00002245-458F-4CDD-8533-B448CD43DBD2   iPhone14,7         Apple
2   3  00012620-313A-4502-9F8D-8DAB7443215B   iPhone14,5         Apple
3   4  000137bc-80de-4bb5-b61d-df7f217a4501     SM-F711N       samsung
4   5  000227D6-B782-4367-91C4-486B76DF9E37   iPhone12,3         Apple

In [11]:
device_properties_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 252380 entries, 0 to 252379
Data columns (total 4 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   id             252380 non-null  int64 
 1   device_id      252380 non-null  object
 2   device_model   252380 non-null  object
 3   device_vendor  252380 non-null  object
dtypes: int64(1), object(3)
memory usage: 7.7+ MB


In [12]:
# 결측치 확인
device_properties_df.isnull().sum()

id               0
device_id        0
device_model     0
device_vendor    0
dtype: int64

In [13]:
# 데이터 중복 확인
device_properties_df[device_properties_df.duplicated()]

Empty DataFrame
Columns: [id, device_id, device_model, device_vendor]
Index: []

## hackle_events

In [14]:
# csv 파일 불러오기
hackle_events_path = r'C:\Users\USER\Documents\codeit_study\codeit_study\HYPE 프로젝트\hackle_events.csv'
hackle_events_df = pd.read_csv(hackle_events_path)

In [15]:
hackle_events_df.head()

event_id       event_datetime  \
0  00000533-3f1c-4b3b-81f1-0c8f35754b4e  2023-07-18 19:40:17   
1  00000716-27e9-4e72-a602-d0ce61784b06  2023-07-18 21:07:24   
2  000007c8-68ce-40e6-9b1e-f0e34e8ff9cc  2023-08-06 20:18:03   
3  00000981-5e2a-4111-993e-4f1891ad9a53  2023-08-05 01:46:10   
4  00000a7a-ba72-4332-b4a9-7910670aaeb2  2023-07-24 15:03:37   

                         event_key                            session_id  \
0                   $session_start          4OzYh3seq3VKytpSn5pvQkZNQii1   
1              click_question_open          8QXy31PQxbW9qLzq0Y1dhR8Ypm52   
2  click_bottom_navigation_profile  6bcea65d-9f40-46fc-888c-700fe707483f   
3                        view_shop          XVYNT6zfhFWqIg9omwg2AHDjTLx2   
4      click_bottom_navigation_lab          XFB2SPiGfjbVhvJ3Q3DBsaT3m2B3   

                                     id item_name page_name  friend_count  \
0  00000533-3f1c-4b3b-81f1-0c8f35754b4e       NaN       NaN           NaN   
1  00000716-27e9-4e72-a602-d0ce61784b06       NaN       NaN          64.0   
2  000007c8-68ce-40e6-9b1e-f0e34e8ff9cc       NaN       NaN          26.0   
3  00000981-5e2a-4111-993e-4f1891ad9a53       NaN       NaN          61.0   
4  00000a7a-ba72-4332-b4a9-7910670aaeb2       NaN       NaN         119.0   

   votes_count  heart_balance  question_id  
0          NaN            NaN          NaN  
1        436.0         4830.0          NaN  
2        174.0         4729.0          NaN  
3         44.0          142.0          NaN  
4        545.0         3287.0          NaN

In [16]:
hackle_events_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11441319 entries, 0 to 11441318
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   event_id        object 
 1   event_datetime  object 
 2   event_key       object 
 3   session_id      object 
 4   id              object 
 5   item_name       object 
 6   page_name       object 
 7   friend_count    float64
 8   votes_count     float64
 9   heart_balance   float64
 10  question_id     float64
dtypes: float64(4), object(7)
memory usage: 960.2+ MB


In [17]:
# 결측치 확인
hackle_events_df.isnull().sum()

event_id                 0
event_datetime           0
event_key                0
session_id               0
id                       0
item_name         11428280
page_name         10652540
friend_count        752556
votes_count         754554
heart_balance       728643
question_id       10991835
dtype: int64

In [18]:
# 데이터 중복 확인
# 중복 없음
hackle_events_df[hackle_events_df.duplicated()]

Empty DataFrame
Columns: [event_id, event_datetime, event_key, session_id, id, item_name, page_name, friend_count, votes_count, heart_balance, question_id]
Index: []

In [19]:
# event_datetime 컬럼을 datetime으로 변환
hackle_events_df['event_datetime'] = pd.to_datetime(hackle_events_df['event_datetime'], errors='coerce')
hackle_events_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11441319 entries, 0 to 11441318
Data columns (total 11 columns):
 #   Column          Dtype         
---  ------          -----         
 0   event_id        object        
 1   event_datetime  datetime64[ns]
 2   event_key       object        
 3   session_id      object        
 4   id              object        
 5   item_name       object        
 6   page_name       object        
 7   friend_count    float64       
 8   votes_count     float64       
 9   heart_balance   float64       
 10  question_id     float64       
dtypes: datetime64[ns](1), float64(4), object(6)
memory usage: 960.2+ MB


## user_properties_df

In [66]:
# csv 파일 불러오기
user_properties_path = r'C:\Users\USER\Documents\codeit_study\codeit_study\HYPE 프로젝트\user_properties.csv'
user_properties_column_names = ['user_id', 'class', 'gender', 'grade', 'school_id']
user_properties_df = pd.read_csv(user_properties_path, header=None, names=user_properties_column_names)

C:\Users\USER\AppData\Local\Temp\ipykernel_15612\2515436470.py:4: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  user_properties_df = pd.read_csv(user_properties_path, header=None, names=user_properties_column_names)


In [67]:
user_properties_df.head()

user_id  class gender  grade  school_id
0  1000000      1      M      1       1885
1  1000009     10      F      2       3869
2  1000012     10      F      1       5091
3  1000013      8      F      2       1743
4  1000015      2      F      3       5078

In [68]:
user_properties_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230819 entries, 0 to 230818
Data columns (total 5 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   user_id    230819 non-null  object
 1   class      230819 non-null  int64 
 2   gender     230819 non-null  object
 3   grade      230819 non-null  int64 
 4   school_id  230819 non-null  int64 
dtypes: int64(3), object(2)
memory usage: 8.8+ MB


In [69]:
# 결측치 확인
user_properties_df.isnull().sum()

user_id      0
class        0
gender       0
grade        0
school_id    0
dtype: int64

In [70]:
# 데이터 중복 확인
user_properties_df[user_properties_df.duplicated()]

Empty DataFrame
Columns: [user_id, class, gender, grade, school_id]
Index: []

## 서비스 내부 데이터.csv 파일 불러오기

유저 이벤트 데이터와 서비스 내부 데이터 확인하기 위한 절차

## account_user_df

In [42]:
# csv 파일 불러오기
accounts_user_path = r'C:\Users\USER\Documents\codeit_study\codeit_study\HYPE 프로젝트\accounts_user.csv'
user_column_names = ['id', 'is_superuser', 'is_staff', 'gender', 'point', 'friend_id_list', 'is_push_on',
                 'created_at', 'block_user_id_list', 'hide_user_id_list', 'ban_status', 'report_count',
                'alarm_count', 'pending_chat', 'pending_votes', 'group_id']
accounts_user_df = pd.read_csv(accounts_user_path, header=None, names=user_column_names)

C:\Users\USER\AppData\Local\Temp\ipykernel_15612\2431421688.py:6: DtypeWarning: Columns (15) have mixed types. Specify dtype option on import or set low_memory=False.
  accounts_user_df = pd.read_csv(accounts_user_path, header=None, names=user_column_names)


In [43]:
accounts_user_df.head()

id  is_superuser  is_staff gender  point  \
0  831956             1         1     \N    600   
1  831962             0         0      F   2248   
2  832151             0         0      M   1519   
3  832340             0         0      F     57   
4  832520             0         0      M   1039   

                                      friend_id_list  is_push_on  \
0  [1292473, 913158, 1488461, 1064695, 1043565, 1...           0   
1  [833025, 832642, 982531, 879496, 838541, 83752...           1   
2  [838785, 982531, 882567, 879496, 838541, 83649...           0   
3  [841345, 982531, 838785, 963714, 882567, 83252...           1   
4  [874050, 849763, 874212, 844297, 838541, 84004...           0   

                   created_at block_user_id_list hide_user_id_list ban_status  \
0  2023-03-29 03:44:14.047130                 []                []          N   
1  2023-03-29 05:18:56.162368                 []                []          N   
2  2023-03-29 12:56:34.989468                 []                []          N   
3  2023-03-29 12:56:35.020790                 []                []          N   
4  2023-03-29 12:56:35.049311                 []                []          N   

   report_count  alarm_count  pending_chat  pending_votes group_id  
0             0            0             0              0       \N  
1           253        40878          5499            110       12  
2             0           37             0             47        1  
3             0           19             0             21        1  
4             0           29             0             15       12

In [44]:
accounts_user_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 677085 entries, 0 to 677084
Data columns (total 16 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   id                  677085 non-null  int64 
 1   is_superuser        677085 non-null  int64 
 2   is_staff            677085 non-null  int64 
 3   gender              677085 non-null  object
 4   point               677085 non-null  int64 
 5   friend_id_list      677085 non-null  object
 6   is_push_on          677085 non-null  int64 
 7   created_at          677085 non-null  object
 8   block_user_id_list  677085 non-null  object
 9   hide_user_id_list   677085 non-null  object
 10  ban_status          677085 non-null  object
 11  report_count        677085 non-null  int64 
 12  alarm_count         677085 non-null  int64 
 13  pending_chat        677085 non-null  int64 
 14  pending_votes       677085 non-null  int64 
 15  group_id            677085 non-null  object
dtypes:

In [45]:
# 데이터 중복 확인
# 중복된 데이터 X
accounts_user_df[accounts_user_df.duplicated()]

Empty DataFrame
Columns: [id, is_superuser, is_staff, gender, point, friend_id_list, is_push_on, created_at, block_user_id_list, hide_user_id_list, ban_status, report_count, alarm_count, pending_chat, pending_votes, group_id]
Index: []

In [46]:
# 결측치 확인
# 결측치 없음
accounts_user_df.isnull().sum()

id                    0
is_superuser          0
is_staff              0
gender                0
point                 0
friend_id_list        0
is_push_on            0
created_at            0
block_user_id_list    0
hide_user_id_list     0
ban_status            0
report_count          0
alarm_count           0
pending_chat          0
pending_votes         0
group_id              0
dtype: int64

In [47]:
# created_at 컬럼의 dtype을 datetime으로 변환
accounts_user_df['created_at'] = pd.to_datetime(accounts_user_df['created_at'], errors='coerce')
accounts_user_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 677085 entries, 0 to 677084
Data columns (total 16 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   id                  677085 non-null  int64         
 1   is_superuser        677085 non-null  int64         
 2   is_staff            677085 non-null  int64         
 3   gender              677085 non-null  object        
 4   point               677085 non-null  int64         
 5   friend_id_list      677085 non-null  object        
 6   is_push_on          677085 non-null  int64         
 7   created_at          677085 non-null  datetime64[ns]
 8   block_user_id_list  677085 non-null  object        
 9   hide_user_id_list   677085 non-null  object        
 10  ban_status          677085 non-null  object        
 11  report_count        677085 non-null  int64         
 12  alarm_count         677085 non-null  int64         
 13  pending_chat        677085 no

# 유저 이벤트 데이터 파일 전처리

### event_id

In [33]:
hackle_events_df.head()

event_id      event_datetime  \
0  00000533-3f1c-4b3b-81f1-0c8f35754b4e 2023-07-18 19:40:17   
1  00000716-27e9-4e72-a602-d0ce61784b06 2023-07-18 21:07:24   
2  000007c8-68ce-40e6-9b1e-f0e34e8ff9cc 2023-08-06 20:18:03   
3  00000981-5e2a-4111-993e-4f1891ad9a53 2023-08-05 01:46:10   
4  00000a7a-ba72-4332-b4a9-7910670aaeb2 2023-07-24 15:03:37   

                         event_key                            session_id  \
0                   $session_start          4OzYh3seq3VKytpSn5pvQkZNQii1   
1              click_question_open          8QXy31PQxbW9qLzq0Y1dhR8Ypm52   
2  click_bottom_navigation_profile  6bcea65d-9f40-46fc-888c-700fe707483f   
3                        view_shop          XVYNT6zfhFWqIg9omwg2AHDjTLx2   
4      click_bottom_navigation_lab          XFB2SPiGfjbVhvJ3Q3DBsaT3m2B3   

                                     id item_name page_name  friend_count  \
0  00000533-3f1c-4b3b-81f1-0c8f35754b4e       NaN       NaN           NaN   
1  00000716-27e9-4e72-a602-d0ce61784b06       NaN       NaN          64.0   
2  000007c8-68ce-40e6-9b1e-f0e34e8ff9cc       NaN       NaN          26.0   
3  00000981-5e2a-4111-993e-4f1891ad9a53       NaN       NaN          61.0   
4  00000a7a-ba72-4332-b4a9-7910670aaeb2       NaN       NaN         119.0   

   votes_count  heart_balance  question_id  
0          NaN            NaN          NaN  
1        436.0         4830.0          NaN  
2        174.0         4729.0          NaN  
3         44.0          142.0          NaN  
4        545.0         3287.0          NaN

In [34]:
hackle_events_df.shape

(11441319, 11)

In [35]:
# hackle_events_df event_id = id 여부 확인
(hackle_events_df['event_id'] == hackle_events_df['id']).value_counts()

True    11441319
Name: count, dtype: int64

In [36]:
# hackle_event_df['id'] 삭제
hackle_events_df.drop(columns='id', inplace=True)

In [37]:
hackle_events_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11441319 entries, 0 to 11441318
Data columns (total 10 columns):
 #   Column          Dtype         
---  ------          -----         
 0   event_id        object        
 1   event_datetime  datetime64[ns]
 2   event_key       object        
 3   session_id      object        
 4   item_name       object        
 5   page_name       object        
 6   friend_count    float64       
 7   votes_count     float64       
 8   heart_balance   float64       
 9   question_id     float64       
dtypes: datetime64[ns](1), float64(4), object(5)
memory usage: 872.9+ MB


In [38]:
hackle_events_df.isna().sum()

event_id                 0
event_datetime           0
event_key                0
session_id               0
item_name         11428280
page_name         10652540
friend_count        752556
votes_count         754554
heart_balance       728643
question_id       10991835
dtype: int64

### user_id

In [39]:
# user_id가 숫자가 아닌 값 찾기
non_numeric_user_ids = hackle_properties_df[~hackle_properties_df['user_id'].apply(lambda x: str(x).isdigit())]
non_numeric_user_ids['user_id'].unique()

array([nan, 'LztzUUFoRxdqTSPgQrX3MAAyNkM2',
       'qLdDlFGK9qObRuGXK20KAGbqzRZ2', ...,
       'RX5m5NGxa5MBJYCo8mLF0eigPu43', 'HYKUXaXF5fbAftyR95G4k9afQ3o2',
       'cpZ6XwYLTkbwDHpBkjp9bPLjVE53'], dtype=object)

In [40]:
non_numeric_user_ids['user_id'].nunique()

96527

In [49]:
non_numeric_user = accounts_user_df[~accounts_user_df['id'].apply(lambda x: str(x).isdigit())]
non_numeric_user['id'].unique()

array([], dtype=int64)

- 서비스 내부 데이터에서 id 값이 숫자가 아닌 값은 없는 것 확인
- 유저 이벤트 데이터에서 id 값이 숫자가 아닌 데이터들은 이상치로 판단, 삭제 처리
    - properties 테이블에서 총 525350개 중 96,527개 값 삭제

In [50]:
hackle_properties_cleaning_df = hackle_properties_df[hackle_properties_df['user_id'].apply(lambda x: str(x).isdigit())]
hackle_properties_cleaning_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 334091 entries, 1 to 525349
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   id           334091 non-null  int64 
 1   session_id   334091 non-null  object
 2   user_id      334091 non-null  object
 3   language     334091 non-null  object
 4   osname       334091 non-null  object
 5   osversion    334091 non-null  object
 6   versionname  334091 non-null  object
 7   device_id    334091 non-null  object
dtypes: int64(1), object(7)
memory usage: 22.9+ MB


### session_id

In [51]:
hackle_events_df['session_id'].value_counts().loc[lambda x: x > 1]

session_id
A40CA2FA-CEB6-4E94-857D-7C229ECC2598    8157
cfa76079-ab4e-4ea1-a7df-38ceaa5ef011    6551
2iO7jwc0YtW0SI8AUjCVoV5gQ1F2            6548
b1b6f602-17ce-4636-9eff-bf0ad5578c5a    5714
6f2a1743-5513-405c-8425-ce11efe267b2    5684
                                        ... 
087D89E0-2FF6-47FA-AABC-3D56E702084F       2
F9Uz4ux3wzXdy0nGqa2WtPavTxB2               2
CD614905-D17A-47F8-8DEB-82777BA4C0B3       2
KBguY11Nm3UIxaYW7BLDNqSMSJ03               2
2bde4dad-4ae4-40b0-863a-691a38aa5f1d       2
Name: count, Length: 250830, dtype: int64

In [52]:
hackle_properties_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525350 entries, 0 to 525349
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   id           525350 non-null  int64 
 1   session_id   525350 non-null  object
 2   user_id      443095 non-null  object
 3   language     525350 non-null  object
 4   osname       525350 non-null  object
 5   osversion    525350 non-null  object
 6   versionname  525350 non-null  object
 7   device_id    525350 non-null  object
dtypes: int64(1), object(7)
memory usage: 32.1+ MB


In [53]:
hackle_properties_df.head()

id                            session_id  user_id language   osname  \
0   1          4OzYh3seq3VKytpSn5pvQkZNQii1      NaN    ko-KR      iOS   
1   2          8QXy31PQxbW9qLzq0Y1dhR8Ypm52  1046711    ko-KR      iOS   
2   3  6bcea65d-9f40-46fc-888c-700fe707483f  1545130       ko  Android   
3   4          XVYNT6zfhFWqIg9omwg2AHDjTLx2  1224793       ko  Android   
4   5          XFB2SPiGfjbVhvJ3Q3DBsaT3m2B3  1329450    ko-US      iOS   

  osversion versionname                             device_id  
0      16.0       2.0.0  590E7C79-CBA0-44D8-8BE3-3C9BFABBBC74  
1    16.5.1       2.0.3  D5417226-F71B-4A9E-A180-CD072F2AB279  
2        13       2.0.5  6bcea65d-9f40-46fc-888c-700fe707483f  
3        13       2.0.5  a05c1595-3e05-434b-8684-218b528bd725  
4    16.5.1       2.0.5  EAC6C0B3-7CE8-40EA-8A91-9977C0BA5EF3

In [54]:
hackle_properties_df['session_id'].duplicated().sum()

np.int64(271734)

In [55]:
hackle_properties_df['session_id'].value_counts().loc[lambda x: x>1]

session_id
040914e1-61ac-40ef-b76a-718066d880dc    9
vheiXNIAkbRNz8OPRfItfkKxm1A2            9
GHbGo2mDG5NRtDGa9Dd8wcT21nS2            8
KxXeIkNRmAVleq2ODKgNwCIFUqE3            8
ShHbFaGn9xg4z8ILJAVjxbm8q672            8
                                       ..
UlSVf7FaWyRzv4l6gSewnIJTYL92            2
yj4Uvtw1NVTf9gEd8xL6EPSZWUC2            2
fLx9o8LfolfY7KQ4ZDzjBrZ1z943            2
3B2A7F23-A212-4FE9-8E9C-27C5A5347489    2
1mnK9pE19gOkqk8Y2qZUGk8LcSx2            2
Name: count, Length: 196324, dtype: int64

In [ ]:
hackle_properties_df[hackle_properties_df['session_id'] == '040914e1-61ac-40ef-b76a-718066d880dc']
# session_id는 동일하지만 user_id가 계속 변동됨

id                            session_id  user_id language  \
9253      9254  040914e1-61ac-40ef-b76a-718066d880dc  1577938       ko   
17152    17153  040914e1-61ac-40ef-b76a-718066d880dc  1577954       ko   
53799    53800  040914e1-61ac-40ef-b76a-718066d880dc  1577938       ko   
101786  101787  040914e1-61ac-40ef-b76a-718066d880dc      NaN       ko   
175568  175569  040914e1-61ac-40ef-b76a-718066d880dc  1577930       ko   
190012  190013  040914e1-61ac-40ef-b76a-718066d880dc   838541       ko   
251638  251639  040914e1-61ac-40ef-b76a-718066d880dc  1577954       ko   
258144  258145  040914e1-61ac-40ef-b76a-718066d880dc      NaN       ko   
315204  315205  040914e1-61ac-40ef-b76a-718066d880dc   849763       ko   

         osname osversion versionname                             device_id  
9253    Android        13       2.0.3  040914e1-61ac-40ef-b76a-718066d880dc  
17152   Android        13       2.0.5  040914e1-61ac-40ef-b76a-718066d880dc  
53799   Android        13       2.0.5  040914e1-61ac-40ef-b76a-718066d880dc  
101786  Android        13       2.0.5  040914e1-61ac-40ef-b76a-718066d880dc  
175568  Android        13       2.0.3  040914e1-61ac-40ef-b76a-718066d880dc  
190012  Android        13       2.0.3  040914e1-61ac-40ef-b76a-718066d880dc  
251638  Android        13       2.0.3  040914e1-61ac-40ef-b76a-718066d880dc  
258144  Android        13       2.0.3  040914e1-61ac-40ef-b76a-718066d880dc  
315204  Android        13       2.0.3  040914e1-61ac-40ef-b76a-718066d880dc

In [ ]:
hackle_properties_df[hackle_properties_df['session_id'] == 'KxXeIkNRmAVleq2ODKgNwCIFUqE3']
# 가끔 user_id와 session_id가 동일하게 나타나는 경우 존재

id                    session_id                       user_id  \
39491    39492  KxXeIkNRmAVleq2ODKgNwCIFUqE3  KxXeIkNRmAVleq2ODKgNwCIFUqE3   
88641    88642  KxXeIkNRmAVleq2ODKgNwCIFUqE3                       1040254   
367607  367608  KxXeIkNRmAVleq2ODKgNwCIFUqE3                       1040254   
413838  413839  KxXeIkNRmAVleq2ODKgNwCIFUqE3                       1040254   
417121  417122  KxXeIkNRmAVleq2ODKgNwCIFUqE3                       1040254   
455887  455888  KxXeIkNRmAVleq2ODKgNwCIFUqE3                       1040254   
511468  511469  KxXeIkNRmAVleq2ODKgNwCIFUqE3                       1040254   
513104  513105  KxXeIkNRmAVleq2ODKgNwCIFUqE3  KxXeIkNRmAVleq2ODKgNwCIFUqE3   

       language osname osversion versionname  \
39491     ko-KR    iOS    15.7.7       2.0.0   
88641     ko-KR    iOS    16.5.1       2.0.3   
367607    ko-KR    iOS      16.6       2.0.5   
413838    ko-KR    iOS    15.7.7       2.0.0   
417121    ko-KR    iOS    15.7.7       2.0.3   
455887    ko-KR    iOS    16.5.1       2.0.5   
511468    ko-KR    iOS    16.5.1       2.0.0   
513104    ko-KR    iOS    16.5.1       2.0.0   

                                   device_id  
39491   28CAA88C-B36D-46EE-B10D-CE989B2176D1  
88641   6C9D8208-B234-4EA0-8CF2-FDC1CE083099  
367607  6C9D8208-B234-4EA0-8CF2-FDC1CE083099  
413838  28CAA88C-B36D-46EE-B10D-CE989B2176D1  
417121  28CAA88C-B36D-46EE-B10D-CE989B2176D1  
455887  6C9D8208-B234-4EA0-8CF2-FDC1CE083099  
511468  6C9D8208-B234-4EA0-8CF2-FDC1CE083099  
513104  6C9D8208-B234-4EA0-8CF2-FDC1CE083099

- session_id = 040914e1-61ac-40ef-b76a-718066d880dc 를 제외하고는 user_id가 nan이거나 문자열인 애를 제외하면 1개의 값으로 중복 제거 될 듯 함
- 같은 session_id, user_id 이지만 다른 device_id
    - 유저 이벤트 데이터를 수집할 때 세션을 생성하는 로직이 누락될 가능성
    - 앱에서 세션을 새로 만들어야 할 시점(재접속, 전환 등)에 새로운 session_id를 만들지 않고 이전 값을 재사용했을 가능성

In [58]:
hackle_properties_cleaning_df = hackle_properties_cleaning_df.drop_duplicates(subset=['session_id', 'user_id'])
hackle_properties_cleaning_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 233668 entries, 1 to 525346
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   id           233668 non-null  int64 
 1   session_id   233668 non-null  object
 2   user_id      233668 non-null  object
 3   language     233668 non-null  object
 4   osname       233668 non-null  object
 5   osversion    233668 non-null  object
 6   versionname  233668 non-null  object
 7   device_id    233668 non-null  object
dtypes: int64(1), object(7)
memory usage: 16.0+ MB


In [59]:
hackle_properties_cleaning_df['session_id'].value_counts().loc[lambda x: x>1]

session_id
040914e1-61ac-40ef-b76a-718066d880dc    5
B941F9F9-CF53-4DAE-A204-75E666B5D277    3
23235867-49A8-4240-8B34-6342357440F9    3
SDCe4wZnO9eBy8aQ1SxAHAIlg7f2            2
b0b04f86-cf77-4cbd-85d9-814ec2bf015e    2
                                       ..
9B0ABDE8-D744-4462-AF6D-876B2DE45CEA    2
D76921C5-5CA5-483E-A32C-F278E49BACC9    2
5c89abcf-a789-4944-8c26-da288ee48d2a    2
21ef715c-d365-405b-a5ad-cdba4fc79ff3    2
96C2C47D-7BC2-4144-8AC6-1FADE599F10A    2
Name: count, Length: 62, dtype: int64

In [60]:
hackle_properties_cleaning_df[hackle_properties_cleaning_df['session_id'] == 'SDCe4wZnO9eBy8aQ1SxAHAIlg7f2']

id                    session_id  user_id language osname  \
99326    99327  SDCe4wZnO9eBy8aQ1SxAHAIlg7f2  1086267    ko-KR    iOS   
147595  147596  SDCe4wZnO9eBy8aQ1SxAHAIlg7f2  1120312    ko-KR    iOS   

       osversion versionname                             device_id  
99326     16.5.1       2.0.5  99DB576B-C1D6-4BCA-9523-CFE02DB3C11B  
147595    16.5.1       2.0.5  99DB576B-C1D6-4BCA-9523-CFE02DB3C11B

In [61]:
hackle_properties_cleaning_df['session_id'].duplicated().sum()

np.int64(67)

- session_id가 같은데 user_id가 다르면서 device_id가 같은 경우
    - 일반 유저로 보기 어려움
    - 관리자 계정일 가능성?
    - 67개의 중복 -> 삭제

In [62]:
hackle_properties_cleaning_df = hackle_properties_cleaning_df.drop_duplicates(subset=['session_id'])
hackle_properties_cleaning_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 233601 entries, 1 to 525346
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   id           233601 non-null  int64 
 1   session_id   233601 non-null  object
 2   user_id      233601 non-null  object
 3   language     233601 non-null  object
 4   osname       233601 non-null  object
 5   osversion    233601 non-null  object
 6   versionname  233601 non-null  object
 7   device_id    233601 non-null  object
dtypes: int64(1), object(7)
memory usage: 16.0+ MB


In [63]:
hackle_properties_cleaning_df['session_id'].duplicated().sum()

np.int64(0)

In [64]:
hackle_properties_cleaning_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 233601 entries, 1 to 525346
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   id           233601 non-null  int64 
 1   session_id   233601 non-null  object
 2   user_id      233601 non-null  object
 3   language     233601 non-null  object
 4   osname       233601 non-null  object
 5   osversion    233601 non-null  object
 6   versionname  233601 non-null  object
 7   device_id    233601 non-null  object
dtypes: int64(1), object(7)
memory usage: 16.0+ MB


## user_properties_df 와 hackle_properties_df 를 merge

In [71]:
user_properties_df.head(2)

user_id  class gender  grade  school_id
0  1000000      1      M      1       1885
1  1000009     10      F      2       3869

In [74]:
hackle_properties_cleaning_df.head(2)

id                            session_id  user_id language   osname  \
1   2          8QXy31PQxbW9qLzq0Y1dhR8Ypm52  1046711    ko-KR      iOS   
2   3  6bcea65d-9f40-46fc-888c-700fe707483f  1545130       ko  Android   

  osversion versionname                             device_id  
1    16.5.1       2.0.3  D5417226-F71B-4A9E-A180-CD072F2AB279  
2        13       2.0.5  6bcea65d-9f40-46fc-888c-700fe707483f

In [75]:
hackle_properties_cleaning_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 233601 entries, 1 to 525346
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   id           233601 non-null  int64 
 1   session_id   233601 non-null  object
 2   user_id      233601 non-null  object
 3   language     233601 non-null  object
 4   osname       233601 non-null  object
 5   osversion    233601 non-null  object
 6   versionname  233601 non-null  object
 7   device_id    233601 non-null  object
dtypes: int64(1), object(7)
memory usage: 16.0+ MB


In [76]:
hackle_properties_cleaning_df['user_id'].duplicated().sum()

np.int64(2797)

- 세션이 다른데, 같은 유저일 수 있어 중복일 수 있음

In [77]:
hackle_user_school_df = pd.merge(user_properties_df, hackle_properties_cleaning_df, on='user_id', how='left')
hackle_user_school_df.head()

user_id  class gender  grade  school_id  id session_id language osname  \
0  1000000      1      M      1       1885 NaN        NaN      NaN    NaN   
1  1000009     10      F      2       3869 NaN        NaN      NaN    NaN   
2  1000012     10      F      1       5091 NaN        NaN      NaN    NaN   
3  1000013      8      F      2       1743 NaN        NaN      NaN    NaN   
4  1000015      2      F      3       5078 NaN        NaN      NaN    NaN   

  osversion versionname device_id  
0       NaN         NaN       NaN  
1       NaN         NaN       NaN  
2       NaN         NaN       NaN  
3       NaN         NaN       NaN  
4       NaN         NaN       NaN

In [78]:
hackle_user_school_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 232148 entries, 0 to 232147
Data columns (total 12 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   user_id      232148 non-null  object 
 1   class        232148 non-null  int64  
 2   gender       232148 non-null  object 
 3   grade        232148 non-null  int64  
 4   school_id    232148 non-null  int64  
 5   id           101037 non-null  float64
 6   session_id   101037 non-null  object 
 7   language     101037 non-null  object 
 8   osname       101037 non-null  object 
 9   osversion    101037 non-null  object 
 10  versionname  101037 non-null  object 
 11  device_id    101037 non-null  object 
dtypes: float64(1), int64(3), object(8)
memory usage: 21.3+ MB


In [79]:
hackle_user_school_df.isna().sum()

user_id             0
class               0
gender              0
grade               0
school_id           0
id             131111
session_id     131111
language       131111
osname         131111
osversion      131111
versionname    131111
device_id      131111
dtype: int64

In [80]:
hackle_user_school_df = hackle_user_school_df.drop(columns=['id', 'language', 'osname', 'osversion', 'versionname'])
hackle_user_school_df.head()

user_id  class gender  grade  school_id session_id device_id
0  1000000      1      M      1       1885        NaN       NaN
1  1000009     10      F      2       3869        NaN       NaN
2  1000012     10      F      1       5091        NaN       NaN
3  1000013      8      F      2       1743        NaN       NaN
4  1000015      2      F      3       5078        NaN       NaN

In [81]:
hackle_user_school_df['school_id'].nunique()

5023

- session_id에 nan값이 많아지기는 하지만, 이건 로그 기록이 없는 학교나 유저라고 간주

## hackle_df 와 hackle_user_school_df 의 merge

In [82]:
hackle_properties_cleaning_df.drop(columns='id', inplace=True)
hackle_properties_cleaning_df.head()

session_id  user_id language   osname osversion  \
1          8QXy31PQxbW9qLzq0Y1dhR8Ypm52  1046711    ko-KR      iOS    16.5.1   
2  6bcea65d-9f40-46fc-888c-700fe707483f  1545130       ko  Android        13   
3          XVYNT6zfhFWqIg9omwg2AHDjTLx2  1224793       ko  Android        13   
4          XFB2SPiGfjbVhvJ3Q3DBsaT3m2B3  1329450    ko-US      iOS    16.5.1   
6  d2b3ca43-4716-4852-b0e2-334848eb66f4  1018001       ko  Android        13   

  versionname                             device_id  
1       2.0.3  D5417226-F71B-4A9E-A180-CD072F2AB279  
2       2.0.5  6bcea65d-9f40-46fc-888c-700fe707483f  
3       2.0.5  a05c1595-3e05-434b-8684-218b528bd725  
4       2.0.5  EAC6C0B3-7CE8-40EA-8A91-9977C0BA5EF3  
6       2.0.5  d2b3ca43-4716-4852-b0e2-334848eb66f4

In [83]:
hackle_events_df['event_id'].duplicated().sum()

np.int64(0)

In [84]:
hackle_events_df.head()

event_id      event_datetime  \
0  00000533-3f1c-4b3b-81f1-0c8f35754b4e 2023-07-18 19:40:17   
1  00000716-27e9-4e72-a602-d0ce61784b06 2023-07-18 21:07:24   
2  000007c8-68ce-40e6-9b1e-f0e34e8ff9cc 2023-08-06 20:18:03   
3  00000981-5e2a-4111-993e-4f1891ad9a53 2023-08-05 01:46:10   
4  00000a7a-ba72-4332-b4a9-7910670aaeb2 2023-07-24 15:03:37   

                         event_key                            session_id  \
0                   $session_start          4OzYh3seq3VKytpSn5pvQkZNQii1   
1              click_question_open          8QXy31PQxbW9qLzq0Y1dhR8Ypm52   
2  click_bottom_navigation_profile  6bcea65d-9f40-46fc-888c-700fe707483f   
3                        view_shop          XVYNT6zfhFWqIg9omwg2AHDjTLx2   
4      click_bottom_navigation_lab          XFB2SPiGfjbVhvJ3Q3DBsaT3m2B3   

  item_name page_name  friend_count  votes_count  heart_balance  question_id  
0       NaN       NaN           NaN          NaN            NaN          NaN  
1       NaN       NaN          64.0        436.0         4830.0          NaN  
2       NaN       NaN          26.0        174.0         4729.0          NaN  
3       NaN       NaN          61.0         44.0          142.0          NaN  
4       NaN       NaN         119.0        545.0         3287.0          NaN

In [85]:
hackle_events_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11441319 entries, 0 to 11441318
Data columns (total 10 columns):
 #   Column          Dtype         
---  ------          -----         
 0   event_id        object        
 1   event_datetime  datetime64[ns]
 2   event_key       object        
 3   session_id      object        
 4   item_name       object        
 5   page_name       object        
 6   friend_count    float64       
 7   votes_count     float64       
 8   heart_balance   float64       
 9   question_id     float64       
dtypes: datetime64[ns](1), float64(4), object(5)
memory usage: 872.9+ MB


In [86]:
hackle_user_school_df.head(2)

user_id  class gender  grade  school_id session_id device_id
0  1000000      1      M      1       1885        NaN       NaN
1  1000009     10      F      2       3869        NaN       NaN

In [87]:
hackle_df = pd.merge(hackle_events_df, hackle_user_school_df, on='session_id', how='left')
hackle_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11441319 entries, 0 to 11441318
Data columns (total 16 columns):
 #   Column          Dtype         
---  ------          -----         
 0   event_id        object        
 1   event_datetime  datetime64[ns]
 2   event_key       object        
 3   session_id      object        
 4   item_name       object        
 5   page_name       object        
 6   friend_count    float64       
 7   votes_count     float64       
 8   heart_balance   float64       
 9   question_id     float64       
 10  user_id         object        
 11  class           float64       
 12  gender          object        
 13  grade           float64       
 14  school_id       float64       
 15  device_id       object        
dtypes: datetime64[ns](1), float64(7), object(8)
memory usage: 1.4+ GB


In [88]:
hackle_df.head()

event_id      event_datetime  \
0  00000533-3f1c-4b3b-81f1-0c8f35754b4e 2023-07-18 19:40:17   
1  00000716-27e9-4e72-a602-d0ce61784b06 2023-07-18 21:07:24   
2  000007c8-68ce-40e6-9b1e-f0e34e8ff9cc 2023-08-06 20:18:03   
3  00000981-5e2a-4111-993e-4f1891ad9a53 2023-08-05 01:46:10   
4  00000a7a-ba72-4332-b4a9-7910670aaeb2 2023-07-24 15:03:37   

                         event_key                            session_id  \
0                   $session_start          4OzYh3seq3VKytpSn5pvQkZNQii1   
1              click_question_open          8QXy31PQxbW9qLzq0Y1dhR8Ypm52   
2  click_bottom_navigation_profile  6bcea65d-9f40-46fc-888c-700fe707483f   
3                        view_shop          XVYNT6zfhFWqIg9omwg2AHDjTLx2   
4      click_bottom_navigation_lab          XFB2SPiGfjbVhvJ3Q3DBsaT3m2B3   

  item_name page_name  friend_count  votes_count  heart_balance  question_id  \
0       NaN       NaN           NaN          NaN            NaN          NaN   
1       NaN       NaN          64.0        436.0         4830.0          NaN   
2       NaN       NaN          26.0        174.0         4729.0          NaN   
3       NaN       NaN          61.0         44.0          142.0          NaN   
4       NaN       NaN         119.0        545.0         3287.0          NaN   

   user_id  class gender  grade  school_id  \
0      NaN    NaN    NaN    NaN        NaN   
1      NaN    NaN    NaN    NaN        NaN   
2  1545130    2.0      M    1.0     1141.0   
3      NaN    NaN    NaN    NaN        NaN   
4      NaN    NaN    NaN    NaN        NaN   

                              device_id  
0                                   NaN  
1                                   NaN  
2  6bcea65d-9f40-46fc-888c-700fe707483f  
3                                   NaN  
4                                   NaN

In [89]:
hackle_df.isna().sum()

event_id                 0
event_datetime           0
event_key                0
session_id               0
item_name         11428280
page_name         10652540
friend_count        752556
votes_count         754554
heart_balance       728643
question_id       10991835
user_id            6291191
class              6291191
gender             6291191
grade              6291191
school_id          6291191
device_id          6291191
dtype: int64

- session 이나 event 가 존재하지만, user_id나 school_id를 인식하지 못하는 정보 다수 발견

In [90]:
hackle_df = hackle_df.dropna(subset=['school_id'])
hackle_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5150128 entries, 2 to 11441318
Data columns (total 16 columns):
 #   Column          Dtype         
---  ------          -----         
 0   event_id        object        
 1   event_datetime  datetime64[ns]
 2   event_key       object        
 3   session_id      object        
 4   item_name       object        
 5   page_name       object        
 6   friend_count    float64       
 7   votes_count     float64       
 8   heart_balance   float64       
 9   question_id     float64       
 10  user_id         object        
 11  class           float64       
 12  gender          object        
 13  grade           float64       
 14  school_id       float64       
 15  device_id       object        
dtypes: datetime64[ns](1), float64(7), object(8)
memory usage: 668.0+ MB


In [91]:
hackle_df.isna().sum()

event_id                0
event_datetime          0
event_key               0
session_id              0
item_name         5144652
page_name         4785820
friend_count       259768
votes_count        260547
heart_balance      251325
question_id       4918710
user_id                 0
class                   0
gender                  0
grade                   0
school_id               0
device_id               0
dtype: int64

In [92]:
hackle_df.head()

event_id      event_datetime  \
2   000007c8-68ce-40e6-9b1e-f0e34e8ff9cc 2023-08-06 20:18:03   
7   00000d23-8055-4534-893b-bf17425ab41e 2023-08-06 12:46:17   
9   000015ae-60e0-495f-a356-9c1aaa18e841 2023-07-21 16:38:37   
11  000016af-25ea-464b-bb59-6f690e9d440b 2023-07-19 00:01:15   
12  0000183e-d7a2-4c47-aa1d-3467dbf41d4f 2023-07-25 19:32:04   

                          event_key                            session_id  \
2   click_bottom_navigation_profile  6bcea65d-9f40-46fc-888c-700fe707483f   
7       click_bottom_navigation_lab  414540BA-1980-4371-BF37-5BFA71158C4D   
9                      view_lab_tap  94860349-d46f-4e98-8505-e96877376cee   
11                       launch_app  496C4656-1079-4284-826B-5D9A8F359ECF   
12               view_questions_tap          cPM87By9qdTxpG7miXyopisBpHU2   

   item_name page_name  friend_count  votes_count  heart_balance  question_id  \
2        NaN       NaN          26.0        174.0         4729.0          NaN   
7        NaN       NaN          45.0        148.0          141.0          NaN   
9        NaN       NaN          21.0        286.0          547.0          NaN   
11       NaN       NaN          70.0        485.0        13364.0          NaN   
12       NaN       NaN          75.0         34.0          626.0          NaN   

    user_id  class gender  grade  school_id  \
2   1545130    2.0      M    1.0     1141.0   
7   1475326    5.0      F    1.0     1204.0   
9   1537177    1.0      M    2.0     5888.0   
11  1556429    1.0      F    2.0      908.0   
12  1517211    6.0      M    2.0     4823.0   

                               device_id  
2   6bcea65d-9f40-46fc-888c-700fe707483f  
7   414540BA-1980-4371-BF37-5BFA71158C4D  
9   94860349-d46f-4e98-8505-e96877376cee  
11  496C4656-1079-4284-826B-5D9A8F359ECF  
12  8d86a4c2-82f1-4ee6-a552-030c724ec709

In [93]:
hackle_df['school_id'].nunique()

4757

In [94]:
event_counts_by_school = hackle_df.groupby('school_id')['event_id'].count().reset_index(name='event_count')
event_counts_by_school.head()

school_id  event_count
0        1.0         4228
1        4.0           45
2        5.0          168
3        6.0           27
4        7.0          234

In [95]:
event_counts_by_school

school_id  event_count
0           1.0         4228
1           4.0           45
2           5.0          168
3           6.0           27
4           7.0          234
...         ...          ...
4752     5959.0           12
4753     5960.0          522
4754     5961.0         3161
4755     5963.0          361
4756     5964.0         1777

[4757 rows x 2 columns]

#### 메인 활동(question) 로그 체크

In [96]:
hackle_df.head(2)

event_id      event_datetime  \
2  000007c8-68ce-40e6-9b1e-f0e34e8ff9cc 2023-08-06 20:18:03   
7  00000d23-8055-4534-893b-bf17425ab41e 2023-08-06 12:46:17   

                         event_key                            session_id  \
2  click_bottom_navigation_profile  6bcea65d-9f40-46fc-888c-700fe707483f   
7      click_bottom_navigation_lab  414540BA-1980-4371-BF37-5BFA71158C4D   

  item_name page_name  friend_count  votes_count  heart_balance  question_id  \
2       NaN       NaN          26.0        174.0         4729.0          NaN   
7       NaN       NaN          45.0        148.0          141.0          NaN   

   user_id  class gender  grade  school_id  \
2  1545130    2.0      M    1.0     1141.0   
7  1475326    5.0      F    1.0     1204.0   

                              device_id  
2  6bcea65d-9f40-46fc-888c-700fe707483f  
7  414540BA-1980-4371-BF37-5BFA71158C4D

In [97]:
hackle_question_df = hackle_df[hackle_df['event_key'].str.contains('question', case=False, na=False)]
hackle_question_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1444597 entries, 12 to 11441315
Data columns (total 16 columns):
 #   Column          Non-Null Count    Dtype         
---  ------          --------------    -----         
 0   event_id        1444597 non-null  object        
 1   event_datetime  1444597 non-null  datetime64[ns]
 2   event_key       1444597 non-null  object        
 3   session_id      1444597 non-null  object        
 4   item_name       0 non-null        object        
 5   page_name       169284 non-null   object        
 6   friend_count    1444595 non-null  float64       
 7   votes_count     1444595 non-null  float64       
 8   heart_balance   1444595 non-null  float64       
 9   question_id     231418 non-null   float64       
 10  user_id         1444597 non-null  object        
 11  class           1444597 non-null  float64       
 12  gender          1444597 non-null  object        
 13  grade           1444597 non-null  float64       
 14  school_id       14445

In [98]:
hackle_question_df['school_id'].nunique()

4670

## 관리자 계정 삭제

- 관리자 계정 831956, 1580627, 1580689

In [99]:
hackle_df[hackle_df['user_id'].isin(['831956', '1580627', '1580689'])]

Empty DataFrame
Columns: [event_id, event_datetime, event_key, session_id, item_name, page_name, friend_count, votes_count, heart_balance, question_id, user_id, class, gender, grade, school_id, device_id]
Index: []

In [100]:
hackle_df[hackle_df['user_id'] == '1580689']

Empty DataFrame
Columns: [event_id, event_datetime, event_key, session_id, item_name, page_name, friend_count, votes_count, heart_balance, question_id, user_id, class, gender, grade, school_id, device_id]
Index: []

In [101]:
hackle_user_school_df[hackle_user_school_df['user_id'].isin(['831956', '1580627', '1580689'])]

Empty DataFrame
Columns: [user_id, class, gender, grade, school_id, session_id, device_id]
Index: []

In [102]:
user_properties_df[user_properties_df['user_id'].isin(['831956', '1580627', '1580689'])]

Empty DataFrame
Columns: [user_id, class, gender, grade, school_id]
Index: []

- 전처리 과정에서 관리자 계정이 삭제되거나, 애초에 유저 이벤트 데이터에 미포함된 것으로 확인

## 유저 수가 적은 학교

In [103]:
hackle_df.head(2)

event_id      event_datetime  \
2  000007c8-68ce-40e6-9b1e-f0e34e8ff9cc 2023-08-06 20:18:03   
7  00000d23-8055-4534-893b-bf17425ab41e 2023-08-06 12:46:17   

                         event_key                            session_id  \
2  click_bottom_navigation_profile  6bcea65d-9f40-46fc-888c-700fe707483f   
7      click_bottom_navigation_lab  414540BA-1980-4371-BF37-5BFA71158C4D   

  item_name page_name  friend_count  votes_count  heart_balance  question_id  \
2       NaN       NaN          26.0        174.0         4729.0          NaN   
7       NaN       NaN          45.0        148.0          141.0          NaN   

   user_id  class gender  grade  school_id  \
2  1545130    2.0      M    1.0     1141.0   
7  1475326    5.0      F    1.0     1204.0   

                              device_id  
2  6bcea65d-9f40-46fc-888c-700fe707483f  
7  414540BA-1980-4371-BF37-5BFA71158C4D

In [104]:
school_user_counts = hackle_df.groupby('school_id')['user_id'].nunique().reset_index()
school_user_counts.columns = ['school_id', 'unique_user_count']

In [105]:
school_user_counts.head()

school_id  unique_user_count
0        1.0                 23
1        4.0                  2
2        5.0                  4
3        6.0                  3
4        7.0                  8

In [107]:
school_user_counts['unique_user_count'].describe()

count    4757.000000
mean       20.960269
std        20.053301
min         1.000000
25%         5.000000
50%        15.000000
75%        31.000000
max       157.000000
Name: unique_user_count, dtype: float64

- 25%가 5명 이하인 학교

In [108]:
(school_user_counts['unique_user_count'] < 3).sum()

np.int64(642)

In [109]:
low_user_schools = school_user_counts[school_user_counts['unique_user_count'] < 3]['school_id']
low_user_schools

1          4.0
6          9.0
8         11.0
12        15.0
17        23.0
         ...  
4745    5950.0
4746    5952.0
4750    5957.0
4752    5959.0
4756    5964.0
Name: school_id, Length: 642, dtype: float64

In [110]:
hackle_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5150128 entries, 2 to 11441318
Data columns (total 16 columns):
 #   Column          Dtype         
---  ------          -----         
 0   event_id        object        
 1   event_datetime  datetime64[ns]
 2   event_key       object        
 3   session_id      object        
 4   item_name       object        
 5   page_name       object        
 6   friend_count    float64       
 7   votes_count     float64       
 8   heart_balance   float64       
 9   question_id     float64       
 10  user_id         object        
 11  class           float64       
 12  gender          object        
 13  grade           float64       
 14  school_id       float64       
 15  device_id       object        
dtypes: datetime64[ns](1), float64(7), object(8)
memory usage: 668.0+ MB


In [111]:
hackle_df['school_id'].nunique()

4757

In [112]:
hackle_df = hackle_df[~hackle_df['school_id'].isin(low_user_schools)]
hackle_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5108810 entries, 2 to 11441318
Data columns (total 16 columns):
 #   Column          Dtype         
---  ------          -----         
 0   event_id        object        
 1   event_datetime  datetime64[ns]
 2   event_key       object        
 3   session_id      object        
 4   item_name       object        
 5   page_name       object        
 6   friend_count    float64       
 7   votes_count     float64       
 8   heart_balance   float64       
 9   question_id     float64       
 10  user_id         object        
 11  class           float64       
 12  gender          object        
 13  grade           float64       
 14  school_id       float64       
 15  device_id       object        
dtypes: datetime64[ns](1), float64(7), object(8)
memory usage: 662.6+ MB


In [113]:
hackle_df['school_id'].nunique()

4115

### 로그가 너무 적은 학교

In [114]:
school_event_counts = hackle_df['school_id'].value_counts().reset_index()
school_event_counts.columns = ['school_id', 'event_count']
school_event_counts.head()

school_id  event_count
0       47.0        66672
1     4867.0        60169
2      400.0        45298
3      215.0        43560
4     3592.0        38374

In [115]:
(school_event_counts['event_count'] <= 20).sum()

np.int64(6)

In [116]:
school_event_counts['event_count'].describe()

count     4115.000000
mean      1241.509113
std       2431.461518
min         15.000000
25%        322.000000
50%        733.000000
75%       1456.000000
max      66672.000000
Name: event_count, dtype: float64

- 유저 수가 적은 학교를 삭제하니, 자연스럽게 삭제됨

## csv 파일로 변환

In [117]:
hackle_df = hackle_df.reset_index(drop=True)

In [118]:
hackle_df.head()

event_id      event_datetime  \
0  000007c8-68ce-40e6-9b1e-f0e34e8ff9cc 2023-08-06 20:18:03   
1  00000d23-8055-4534-893b-bf17425ab41e 2023-08-06 12:46:17   
2  000015ae-60e0-495f-a356-9c1aaa18e841 2023-07-21 16:38:37   
3  000016af-25ea-464b-bb59-6f690e9d440b 2023-07-19 00:01:15   
4  0000183e-d7a2-4c47-aa1d-3467dbf41d4f 2023-07-25 19:32:04   

                         event_key                            session_id  \
0  click_bottom_navigation_profile  6bcea65d-9f40-46fc-888c-700fe707483f   
1      click_bottom_navigation_lab  414540BA-1980-4371-BF37-5BFA71158C4D   
2                     view_lab_tap  94860349-d46f-4e98-8505-e96877376cee   
3                       launch_app  496C4656-1079-4284-826B-5D9A8F359ECF   
4               view_questions_tap          cPM87By9qdTxpG7miXyopisBpHU2   

  item_name page_name  friend_count  votes_count  heart_balance  question_id  \
0       NaN       NaN          26.0        174.0         4729.0          NaN   
1       NaN       NaN          45.0        148.0          141.0          NaN   
2       NaN       NaN          21.0        286.0          547.0          NaN   
3       NaN       NaN          70.0        485.0        13364.0          NaN   
4       NaN       NaN          75.0         34.0          626.0          NaN   

   user_id  class gender  grade  school_id  \
0  1545130    2.0      M    1.0     1141.0   
1  1475326    5.0      F    1.0     1204.0   
2  1537177    1.0      M    2.0     5888.0   
3  1556429    1.0      F    2.0      908.0   
4  1517211    6.0      M    2.0     4823.0   

                              device_id  
0  6bcea65d-9f40-46fc-888c-700fe707483f  
1  414540BA-1980-4371-BF37-5BFA71158C4D  
2  94860349-d46f-4e98-8505-e96877376cee  
3  496C4656-1079-4284-826B-5D9A8F359ECF  
4  8d86a4c2-82f1-4ee6-a552-030c724ec709

In [ ]:
hackle_df.to_csv('hackle_df.csv', index=False)

# 정리

1. hackle_event_df
    - event_id 와 id 컬럼이 중복되어 삭제
    - user_id가 숫자가 아닌 데이터 삭제
2. hackle_properties_df
    - user_id, session_id 삭제
    - 한 session_id에서 user_id가 여러 개 존재하면서 device_id가 같은 경우 삭제
    - session_id, user_id는 같지만 device_id가 다른 데이터 삭제
3. merge 과정
    - user_properties_df 와 hackle_properties_df 를 user_id 기준으로 left join (hackle_user_school_df)
    - hackle_events_df 와 hackle_user_school_df 를 session_id 기준으로 left join (hackle_df)
4. merge 이후 데이터 정제
    - 학교 정보가 없는 데이터 삭제
    - 관리자 계정 데이터 삭제
    - 유저 수 3명 미만인 학교에 해당하는 데이터 삭제